src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php

```diff
@@ -905,22 +911,65 @@ private function createRequestMatcher(ContainerBuilder $container, string $path
             }
         }
 
-        $id = '.security.request_matcher.'.ContainerBuilder::hash([$path, $host, $port, $methods, $ips, $attributes]);
+        $id = '.security.request_matcher.'.ContainerBuilder::hash([ChainRequestMatcher::class, $path, $host, $port, $methods, $ips, $attributes]);
 
         if (isset($this->requestMatchers[$id])) {
             return $this->requestMatchers[$id];
         }
 
-        // only add arguments that are necessary
-        $arguments = [$path, $host, $methods, $ips, $attributes, null, $port];
-        while (\count($arguments) > 0 && !end($arguments)) {
-            array_pop($arguments);
+        $arguments = [];
+        if ($methods && !$container->hasDefinition($lid = '.security.request_matcher.'.ContainerBuilder::hash([MethodRequestMatcher::class, $methods]))) {
+            $container->register($lid, MethodRequestMatcher::class)
+                ->setPublic(false)
+                ->setArguments([$methods])
+            ;
+            $arguments[] = new Reference($lid);
+        }
+
+        if ($path && !$container->hasDefinition($lid = '.security.request_matcher.'.ContainerBuilder::hash([PathRequestMatcher::class, $path]))) {
+            $container->register($lid, PathRequestMatcher::class)
+                ->setPublic(false)
+                ->setArguments([$path])
+            ;
+            $arguments[] = new Reference($lid);
+        }
+
+        if ($host && !$container->hasDefinition($lid = '.security.request_matcher.'.ContainerBuilder::hash([HostRequestMatcher::class, $host]))) {
+            $container->register($lid, HostRequestMatcher::class)
+                ->setPublic(false)
+                ->setArguments([$host])
+            ;
+            $arguments[] = new Reference($lid);
+        }
+
+        if ($ips && !$container->hasDefinition($lid = '.security.request_matcher.'.ContainerBuilder::hash([IpsRequestMatcher::class, $ips]))) {
+            $container->register($lid, IpsRequestMatcher::class)
+                ->setPublic(false)
+                ->setArguments([$ips])
+            ;
+            $arguments[] = new Reference($lid);
+        }
+
+        if ($attributes && !$container->hasDefinition($lid = '.security.request_matcher.'.ContainerBuilder::hash([AttributesRequestMatcher::class, $attributes]))) {
+            $container->register($lid, AttributesRequestMatcher::class)
+                ->setPublic(false)
+                ->setArguments([$attributes])
+            ;
+            $arguments[] = new Reference($lid);
+        }
+
+        if ($port && !$container->hasDefinition($lid = '.security.request_matcher.'.ContainerBuilder::hash([PortRequestMatcher::class, $port]))) {
+            $container->register($lid, PortRequestMatcher::class)
+                ->setPublic(false)
```

Comment:
`setPublic(false)` should not be necessary since `false` is the default (same for all definitions)